src/net/transport.rs

```diff
@@ -30,115 +36,146 @@ impl HttpTransport {
-                let mut ongoing_queries = self.ongoing_queries.lock().unwrap();
-                match ongoing_queries.entry(query_id) {
-                    Entry::Occupied(_) => {
-                        panic!("attempted to subscribe to commands for query id {}, but there is already a previous subscriber", query_id.as_ref())
-                    }
-                    Entry::Vacant(entry) => {
-                        let (tx, rx) = mpsc::channel(1);
-                        entry.insert(tx);
-                        ReceiverStream::new(rx)
-                    }
-                }
+    async fn send<
+        D: Stream<Item = Vec<u8>> + Send + 'static,
+        Q: QueryIdBinding,
+        S: StepBinding,
+        R: RouteParams<RouteId, Q, S>,
+    >(
+        &self,
+        dest: HelperIdentity,
+        route: R,
+        data: D,
+    ) -> Result<(), TransportError>
+    where
+        Option<QueryId>: From<Q>,
+        Option<Step>: From<S>,
+    {
+        let route_id = route.resource_identifier();
+        match route_id {
+            RouteId::Records => {
+                // TODO(600): These fallible extractions aren't really necessary.
+                let query_id = extract_route_param(route.query_id(), "query_id")?;
+                let step = extract_route_param(route.step(), "step")?;
+                let resp_future = self.clients[dest]
+                    .step(dest, query_id, &step, data)
+                    .await
+                    .map_err(|_e| {
+                        io::Error::new(io::ErrorKind::ConnectionAborted, "channel closed")
+                    })?;
+                tokio::spawn(async move {
+                    resp_future
+                        .map_err(Into::into)
+                        .and_then(MpcHelperClient::resp_ok)
+                        .await
+                        .expect("failed to stream records");
+                });
+                // TODO(600): We need to do something better than panic if there is an error sending the
+                // data. Note, also, that the caller of this function (`GatewayBase::get_sender`)
+                // currently panics on errors.
+                Ok(())
+            }
+            RouteId::PrepareQuery => {
+                let req = serde_json::from_str(route.extra().borrow()).unwrap();
+                self.clients[dest]
+                    .prepare_query(self.identity, req)
+                    .await
+                    .map_err(|e| TransportError::Rejected {
+                        dest,
+                        inner: Box::new(e),
+                    })
+            }
+            RouteId::ReceiveQuery => {
+                // TODO(600): Can we eliminate cases for client requests from this function
+                // entirely?
+                Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "can't send ReceiveQuery to another helper",
+                )
+                .into())
```

Comment:
An `unreachable!()` should do here.

Reply:
In that case, do you also want to change the the formerly `expect("TODO")` cases to panic if somebody tries to send records (step data) without specifying query or step ID?